### ISSUE TITLE
DJIAoaControllerActivity crashing when attempting to log Intent Action

### ISSUE THREAD

**Issue by jtuchek** [dji-sdk/Mobile-SDK-Android#39]
Android Devices: Nexus 9, Nexus 6P
DJI Product: P3 Controller and Phantom 3 Professional
Android Studio: v2.2
Android Version: 7.0

When plugging in a USB device, DJIAoAControllerActivity causes the application to crash. It appears that the Intent->Action isn't available and somewhere in the logging process it throws a NullPointerException.

It's declaration in the AndroidManifest also has a default launchMode, including a snippet:
`<activity
            android:name="dji.sdk.sdkmanager.DJIAoaControllerActivity"
            android:theme="@android:style/Theme.Translucent">
            <intent-filter>
                <action android:name="android.hardware.usb.action.USB_ACCESSORY_ATTACHED" />
            </intent-filter>

```
        <meta-data
            android:name="android.hardware.usb.action.USB_ACCESSORY_ATTACHED"
            android:resource="@xml/usb_accessory_filter" />
    </activity>`
```

The issue weird because it should only launch the Activity when the filter is satisfied, but once it is started it doesn't seem to retain Action - so even if `LOG.E` didn't throw an NPE it would fail the check and _not_ broadcast the DJISDKManager.USB_ACCESSORY_ATTACHED Intent.

Stack trace
`10-06 08:50:33.933 31827-31827/com.botlink.app E/AndroidRuntime: FATAL EXCEPTION: main
                                                                 Process: com.botlink.app, PID: 31827
                                                                 java.lang.RuntimeException: Unable to start activity ComponentInfo{com.botlink.app/dji.sdk.sdkmanager.DJIAoaControllerActivity}: java.lang.NullPointerException
                                                                     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2646)
                                                                     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2707)
                                                                     at android.app.ActivityThread.-wrap12(ActivityThread.java)
                                                                     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1460)
                                                                     at android.os.Handler.dispatchMessage(Handler.java:102)
                                                                     at android.os.Looper.loop(Looper.java:154)
                                                                     at android.app.ActivityThread.main(ActivityThread.java:6077)
                                                                     at java.lang.reflect.Method.invoke(Native Method)
                                                                     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865)
                                                                     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755)
                                                                  Caused by: java.lang.NullPointerException
                                                                     at java.io.File.<init>(File.java:262)
                                                                     at dji.log.DJILogHelper.saveLogInfo2File(Unknown Source)
                                                                     at dji.log.DJILogHelper.LOGE(Unknown Source)
                                                                     at dji.sdk.sdkmanager.DJIAoaControllerActivity.onCreate(Unknown Source)
                                                                     at android.app.Activity.performCreate(Activity.java:6664)
                                                                     at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1118)
                                                                     at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2599)
                                                                     at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2707) 
                                                                     at android.app.ActivityThread.-wrap12(ActivityThread.java) 
                                                                     at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1460) 
                                                                     at android.os.Handler.dispatchMessage(Handler.java:102) 
                                                                     at android.os.Looper.loop(Looper.java:154) 
                                                                     at android.app.ActivityThread.main(ActivityThread.java:6077) 
                                                                     at java.lang.reflect.Method.invoke(Native Method) 
                                                                     at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:865) 
                                                                     at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:755) `

**Comment by jtuchek** [comment]
It may be worth mentioning I can't reproduce this issue 100% of the time - maybe 40% of the time. It's happened around 5-6 times this morning in the last 2 hours in the office.

**Comment by kenargo** [comment]
You didn’t mention which SDK version?  If you haven’t tried with 3.3 you might want to see if that works/fails.

**Comment by jtuchek** [comment]
Sorry about that - I've been on SDK v3.3 for the last few weeks, even checked the result of DJISDKManager->getSDKVersion(...) to be sure.

**Comment by jtuchek** [comment]
Update - I am able to reproduce it consistently. It's not the Intent Action that is the problem - it's the internal `LOG` class.

If `DJIAoAControllerActivity` receives USB_ACCESSORY_CONNECTED event AND the SDK has not yet been initialized it will crash when attempting to invoke the logger.

Would it be possible to put those actions behind a registration check - such as
`    protected void onCreate(Bundle var1) {
        super.onCreate(var1);
        this.setContentView(new View(this));

        Intent var2 = this.getIntent();
        if(var2 != null && registered()) {
            String var3 = var2.getAction();
            DJILogHelper.getInstance().LOGE("", "action=" + var3, true, true);
            if(var3 == "android.hardware.usb.action.USB_ACCESSORY_ATTACHED") {
                Intent var4 = new Intent();
                var4.setAction(DJISDKManager.USB_ACCESSORY_ATTACHED);
                this.sendBroadcast(var4);
                DJILogHelper.getInstance().LOGE("", "action=send", true, true);
            }
        }

        this.finish();
    }

    protected void onDestroy() {
        super.onDestroy();
        if(registered) {
          DJILogHelper.getInstance().LOGE("", "DJIAoaActivity onDestroy", true, true);
        }
    }

    protected boolean registered() {
      return DJISDKManger.getInstance().hasSDKRegistered();
    }`

If you're an application that supports additional USB peripherals this can be an unfriendly API (if you're lazily initializing the SDK or it has failed for a multitude of other reasons such as internet connectivity, call failure, etc) - especially since there's no reasonable defense an implementer would have against this.

**Comment by jtuchek** [comment]
I ended up writing my own Activity and forwarding the `DJISDKManager.USB_ACCESSORY_ATTACHED` by hand - as subclassing isn't an option because optionally invoking super will crash anyways if an Activity is not correctly prepared.

**Comment by kenargo** [comment]
Would you care to share the key parts of your activity so everyone can benefit from your investigation?

**Comment by jtuchek** [comment]
@kenargo it is functionally equivelant to what is 2 comments up - it queries the DJISDKManager to see if registration has already succeeded (to initialize components).

if SDKManager not registered, tell activity to finish - bypassing blocks to call things like Log that have not been initialized.

**Comment by kenargo** [comment]
Is this only an issues with 3.3 sdk?

**Comment by jtuchek** [comment]
@kenargo to be honest, I'm not sure - support window at the moment is exclusively 3.3, so no prior versions out in the wild.

Since the actual source isn't available on GitHub I'd have to download all prior releases, test, etc - **insert stuff about legalities and decompilation, etc** - obviously since it's not really open source...I can't exactly push up a patch or inspect any prior versions without conscent or taking a finer read at the SDK EULA.

**Comment by kenargo** [comment]
Yes, I understand but I can say with certainty that I wouldn't release any public app using 3.3.  I've found it to be randomly unstable, crashes on many devices and some devices that worked before are now broken.

**Comment by jtuchek** [comment]
I can't say that I've had the same experience yet - although we have a fairly low range of Android SDK support (18-24) and a limited usage of the surface API: primarily MissionManager, Aircraft, and marginal parts of the Camera API.

I'd be curious to see how our usage aligns - maybe it'd be more appropriate to move this to a different thread or communication. I wish the Mobile-SDK has a gitter.im channel setup like they do for the Onboard-SDK.

**Comment by kenargo** [comment]
Try any Samsung Tab units, the tab 3, 4, E all crash.   My S7 crashes if I have the service entry in the manifest, it's that easy.  In fact with my S7 I don't even need to turn on the aircraft, just a main activity and the service references is good enough to cause the crash. 

I thought I had the issues resolved by not including the service reference, which worked for most devices but then the Tab devices crashed so I've rolled back my app to use 2.2.1 and all issues when away.

**Comment by dji-dev** [comment]
Hi @kenargo @jtuchek, thanks for your feedbacks. One question for you two, have you moved the SDK initialization code to other places rather than in the application (Like in MainActivity.java)? Please double check this.

Also, it could be a bug in SDK 3.3 (DJILogHelper's problem as @jtuchek said), we will improve it in the next SDK version release. For now, we provide a hotfix for you, please create a new Activity class named "DJIAoaControllerCompatActivity.java" and replace the code of it with the followings:

~~~java
public class DJIAoaControllerCompatActivity extends Activity {

    /**
     * Constructor.
     */
    public DJIAoaControllerCompatActivity() {
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(new View(this));

        Intent aoaIntent = getIntent();
        if (aoaIntent!=null) {
            String action = aoaIntent.getAction();
            if (action==UsbManager.ACTION_USB_ACCESSORY_ATTACHED) {
                Intent attachedIntent=new Intent();
                attachedIntent.setAction(DJISDKManager.USB_ACCESSORY_ATTACHED);
                sendBroadcast(attachedIntent);
            }
        }

        finish();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
}
~~~

Then update the original **DJIAoaControllerActivity** path with this new activity file's path in the following position in **AndroidManifest.xml** file:

<img width="822" alt="screen shot 2016-10-08 at 4 13 06 pm" src="https://cloud.githubusercontent.com/assets/11869194/19211598/e1e94d4a-8d72-11e6-96c1-8471ad757588.png">

Delete the app in your Android device, do a clean build and try again. Hope this help!

**Comment by kenargo** [comment]
I'll try in the morning.   Will this help with the issues around the crashed with geo or just the initial crash/log issue?

**Comment by jtuchek** [comment]
@dji-dev initialization in the Application class - it is at the point of connection.

It's essentially a chained Rx call - for a number of reasons

* it starts up a large number of services and there are native bindings, so then I can provide an error instead of crashing
* Initialization is coupled to registration, so without internet the setup fails (although resources are initialized)

The chain is
* check if registered, if yes it's complete - safe to access DJISDKManager
* else, call initialization. this is only allowed to be invoked once and caches the result
* else, initialization call has been consumed - just invoke registration.

I'm sure that is why the internal logger is failing, receiving the accessory connection event before initialized - although I thought it seemed a bit harsh to crash because of an external analytics system.

I could rework that into the application - except on newer versions of Android it would not complete do to some deep internals needing read and write external storage permission which were not yet accepted by the user and can only be invoked from an activity context.

**Comment by dji-dev** [comment]
@kenargo The hotfix is just for the initial crash/log issue.

@jtuchek 
1. First of all, the SDK initial must put in the Application class. We cannot guarantee the SDK works well when you put the SDK initial code in other places, and we don't have a plan to change this.
2. The DJIAoaControllerActivity is necessary, because it is the only way to detect the USB Plug-In status, also it is limited by Android AOA Connection, you can check more details from the following link: <http://stackoverflow.com/questions/11098332/usb-accessory-catch-usb-accessory-attached-by-intent-filter-of-service>
3. I can't understand what you said in the last paragraph:
~~~ 
except on newer versions of Android it would not complete do to some deep internals needing read and write external storage permission which were not yet accepted by the user and can only be invoked from an activity context."
~~~
Could you explain in more details or in Chinese (If you know Chinese). Thanks!

**Comment by kenargo** [comment]
I have only code in app, same as demo.

**Comment by dji-dev** [comment]
@kenargo does the hotfix help you to solve the issue?

**Comment by jtuchek** [comment]
@dji-dev 

I can make an adjustment #1 - should be a simple fix. 
Totally understand number #2 - we do the same thing for USB_DEVICE_CONNECTED captures.

Unfortunately I have not learned any Chinese yet - I'll try and provide more a more detailed/simple explanation.

An Android SDK 23+ there seems to be an issue w/ runtime permissions and calling DJISDKManager.initialize(...)

From testing I believe it is related to the "android.permission.READ_PHONE_STATE" permission which is not granted automatically. So when this is called the DJILogHelper attempts to write a few files and hits FileNotFoundExceptions and ultimately sends this out to Logcat "Exception is: getDeviceId: Neither user 10245 nor current process has android.permission.READ_PHONE_STATE."

The internal register network call will complete successfully, but the callback for initialization will never be returned - it just waits infinitely (or at the 3+ minutes I watched.) So the callback doesn't return an DJIError and even the the register call is completed when calling DJISDKManager.isSDKRegistered() it will return false - after this happens there seems to be no amount of DJISDKManager.registerApp() calls that will make this correct or safely initialize.

That's why I did not have the initialization in the application, because without first having the proper permissions - it would never be properly initialized.

**Comment by jtuchek** [comment]
@dji-dev I do have a coworker that knows Chinese - if the explanation above doesn't help I can get his assistance translating.

**Comment by dji-dev** [comment]
@jtuchek thanks for your feedbacks. Fully understand your questions. Actually, we also found this Android SDK version 23+ issue and will update the sample code of the Android SDK in a few days with a hotfix. 

For now, If you are urgent, you can try to modify the `compileSdkVersion` and `targetSdkVersion` back to **22** in the "build.gradle" file to temporarily fix this issue:

![screen shot 2016-10-11 at 11 40 59 am](https://cloud.githubusercontent.com/assets/11869194/19258102/c07a5df2-8fa7-11e6-9fc8-842bcc7b52bf.png)

Hope this help!

**Comment by jtuchek** [comment]
@dji-dev thanks for the update.

**Comment by MustafaMunir** [comment]
@dji-dev is this issue fixed with the new SDK 3.4 ? if not, it would be helpful if you can point me towards the hotfix in the sample code for this issue.

Thanks

**Comment by oliverou** [comment]
Hi @MustafaMunir, in SDK 3.4, we have removed the DJILogHelper, you can give it a try and check if the issue still exist. 

Also,  we provide a hotfix for you, please create a new Activity class named "DJIAoaControllerCompatActivity.java" and replace the code of it with the followings:

~~~java
public class DJIAoaControllerCompatActivity extends Activity {

    /**
     * Constructor.
     */
    public DJIAoaControllerCompatActivity() {
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(new View(this));

        Intent aoaIntent = getIntent();
        if (aoaIntent!=null) {
            String action = aoaIntent.getAction();
            if (action==UsbManager.ACTION_USB_ACCESSORY_ATTACHED) {
                Intent attachedIntent=new Intent();
                attachedIntent.setAction(DJISDKManager.USB_ACCESSORY_ATTACHED);
                sendBroadcast(attachedIntent);
            }
        }

        finish();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
    }
}
~~~

Then update the original DJIAoaControllerActivity path with this new activity file's path in the following position in AndroidManifest.xml file:

<img width="822" alt="screen shot 2016-10-08 at 4 13 06 pm" src="https://cloud.githubusercontent.com/assets/11869194/19211598/e1e94d4a-8d72-11e6-96c1-8471ad757588.png">

Delete the app in your Android device, do a clean build and try again. Hope this help!

**Comment by jtuchek** [comment]
@dji-dev @oliverou @MustafaMunir appears to be fixed in 3.4 - thank you!
